### PR TITLE
Allow traversing into serialized meta/option field values

### PR DESCRIPTION
**Pull request by JoryHogeveen**
Fixes: #5603 

## How Has This Been Tested?

Might be good to add unit tests to make sure this feature works correctly and nothing else breaks with it.

## ChangeLog

- Feature: Allow traversing into serialized meta/option field values.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.
- [x] My code includes PHP Unit Tests (if applicable)

### CODE REVIEW

File: classes/Pods.php [hunk 1]
```diff
@@ -1679,7 +1679,9 @@ public function field( $name, $single = null, $raw = false ) {
 
 									foreach ( $data as $item_id => $item ) {
 										// $field is 123x123, needs to be _src.123x123
-										$full_field = implode( '.', array_splice( $params->traverse, $key ) );
+										$traverse_fields = array_splice( $params->traverse, $key );
+										$full_field      = implode( '.', $traverse_fields );
+										$maybe_traverse  = false;
```
Comment (sc0ttkclark) [inline classes/Pods.php]:
```suggestion
										$traverse_fields    = array_splice( $params->traverse, $key );
										$full_field         = implode( '.', $traverse_fields );
										$maybe_traverse     = false;
										$maybe_traverse_val = null;
```

File: classes/Pods.php [hunk 2]
```diff
@@ -1807,10 +1810,24 @@ public function field( $name, $single = null, $raw = false ) {
...
+											$maybe_traverse     = true;
 										}//end if
+
+										if ( isset( $maybe_traverse_val ) ) {
+											if ( $maybe_traverse && is_array( $maybe_traverse_val ) && 1 < count( $traverse_fields ) ) {
+												array_shift( $traverse_fields );
+												foreach ( $traverse_fields as $field ) {
+													if ( isset( $maybe_traverse_val[ $field ] ) ) {
+														$maybe_traverse_val = $maybe_traverse_val[ $field ];
+													}
```
Comment (sc0ttkclark) [inline classes/Pods.php]:
```suggestion
													if ( ! isset( $maybe_traverse_val[ $field ] ) ) {
														break;
													}

													$maybe_traverse_val = $maybe_traverse_val[ $field ];
```

File: classes/Pods.php [hunk 3]
```diff
@@ -1807,10 +1810,26 @@ public function field( $name, $single = null, $raw = false ) {
...
+											$maybe_traverse_val = get_option( $object . '_' . $field );
+											$maybe_traverse     = true;
 										}//end if
+
+										if ( isset( $maybe_traverse_val ) ) {
+											if ( $maybe_traverse && is_array( $maybe_traverse_val ) && 1 < count( $traverse_fields ) ) {
+												array_shift( $traverse_fields );
+												foreach ( $traverse_fields as $field ) {
+													if ( ! isset( $maybe_traverse_val[ $field ] ) ) {
+														break;
```
Comment (sc0ttkclark) [inline classes/Pods.php]: @JoryHogeveen If we find that the traverse field is not set in the value, should we return a null or empty string for the `$maybe_traverse_val`?

```suggestion
														$maybe_traverse_val = null;

														break;
```

**Comment by JoryHogeveen** [comment]
Might be nice if @jimtrue could take a look at this one. My tests were successful

**Comment by jimtrue** [comment]
Did you guys actually create a test case for this? ie Package Export? Be happy to test it, but I can verify it doesn't break anything in the each/etc. of Templates. Can't validate it actually works with out a good example. I'm guessing the traversal is only supported in the straight traversal, not within each, right?

**Comment by JoryHogeveen** [comment]
See the related issue for a magictag test with image fields.
If you create an image relationship you can fetch values from the serialized stored metadata (`_wp_attachment_metadata`).
Same for other metadata. If it's stored serialized you can now select data from that array.

The reason for this PR was a wp.org topic where the user wanted to get the metadata from an image.

**Comment by jimtrue** [comment]
@JoryHogeveen Remember, I'm volunteering to test. I mostly test to make it sure it doesn't break existing templates handling. If you want me to test a specific one like this, I really need a package export and template code so I can see what you're doing and then apply that to an existing template to make sure it also works within `each` loops, etc.

File: classes/Pods.php [hunk 4]
```diff
@@ -1675,7 +1659,9 @@ public function field( $name, $single = null, $raw = false ) {
 
 									foreach ( $data as $item_id => $item ) {
 										// $field is 123x123, needs to be _src.123x123
-										$full_field = implode( '.', array_splice( $params->traverse, $key ) );
+										$traverse_fields    = array_splice( $params->traverse, $key );
```
Comment (tr1b0t) [inline classes/Pods.php]: :warning: Equals sign not aligned with surrounding assignments; expected 1 space but found 4 spaces

<sub>hash: eb5777b2ce805d2ecfca87006cbd33cd</sub>

File: classes/Pods.php [hunk 5]
```diff
@@ -1675,7 +1659,9 @@ public function field( $name, $single = null, $raw = false ) {
...
+										$full_field         = implode( '.', $traverse_fields );
```
Comment (tr1b0t) [inline classes/Pods.php]: :warning: Equals sign not aligned with surrounding assignments; expected 6 spaces but found 9 spaces

<sub>hash: 405a27f6be2b68bf6e8e49019505b456</sub>

File: includes/data.php [hunk 6]
```diff
@@ -265,6 +265,43 @@ function pods_unslash( $input ) {
 
 }
 
+/**
+ * Traverse an array or object by array values order or a string (name.name.name).
+ *
+ * @since 2.7.18
+ *
+ * @param array|string $traverse
```
Comment (tr1b0t) [inline includes/data.php]: :no_entry_sign: Missing parameter comment

<sub>hash: a11127554e6a19fa77e948f33b0ce194</sub>

File: includes/data.php [hunk 7]
```diff
@@ -265,6 +265,43 @@ function pods_unslash( $input ) {
...
+ * @param array|object $value
```
Comment (tr1b0t) [inline includes/data.php]: :no_entry_sign: Missing parameter comment

<sub>hash: 8b2214f3dc3b6497ac266d69a6541333</sub>

File: classes/Pods.php [hunk 8]
```diff
@@ -1109,27 +1109,16 @@ public function field( $name, $single = null, $raw = false ) {
...
-											$value = null;
-										}
+								$value      = pods_v( implode( '.', $traverse_names ), $attachment, null );
+
+								if ( null === $value ) {
+									// Start traversal though object property or metadata.
+									$name_key = array_shift( $traverse_names );
+									$value    = pods_v( $name_key, $attachment, null );
+									if ( null === $value ) {
+										$value    = get_post_meta( $attachment_id, $name_key, true );
```
Comment (tr1b0t) [inline classes/Pods.php]: :warning: Equals sign not aligned correctly; expected 1 space but found 4 spaces

<sub>hash: 496321c1fdfbc85b36973cdecbce3f85</sub>

File: classes/Pods.php [hunk 9]
```diff
@@ -1109,27 +1109,16 @@ public function field( $name, $single = null, $raw = false ) {
...
 									}
+									$value    = pods_traverse( $traverse_names, $value );
```
Comment (tr1b0t) [inline classes/Pods.php]: :warning: Equals sign not aligned correctly; expected 1 space but found 4 spaces

<sub>hash: d27a4c84406aa57615071edcc3ddc7b1</sub>

File: classes/Pods.php [hunk 10]
```diff
@@ -1109,27 +1109,16 @@ public function field( $name, $single = null, $raw = false ) {
...
+								$value      = pods_v( implode( '.', $traverse_names ), $attachment, null );
+
+								if ( null === $value ) {
+									// Start traversal though object property or metadata.
+									$name_key = array_shift( $traverse_names );
+									$value    = pods_v( $name_key, $attachment, null );
+									if ( null === $value ) {
+										$value = get_post_meta( $attachment_id, $name_key, true );
 									}
+									$value = pods_traverse( $traverse_names, $value );
```
Comment (sc0ttkclark) [inline classes/Pods.php:1121]:
```suggestion

									$value = pods_traverse( $traverse_names, $value );
```

File: classes/Pods.php [hunk 11]
```diff
@@ -1109,27 +1109,16 @@ public function field( $name, $single = null, $raw = false ) {
...
-										} else {
-											$value = null;
-										}
+								$value      = pods_v( implode( '.', $traverse_names ), $attachment, null );
+
+								if ( null === $value ) {
+									// Start traversal though object property or metadata.
+									$name_key = array_shift( $traverse_names );
+									$value    = pods_v( $name_key, $attachment, null );
+									if ( null === $value ) {
```
Comment (sc0ttkclark) [inline classes/Pods.php:1118]:
```suggestion

									if ( null === $value ) {
```

File: includes/data.php [hunk 12]
```diff
@@ -310,6 +310,43 @@ function pods_trim( $input, $charlist = null, $lr = null ) {
...
+ * @param array|string|int $traverse The traversal names/keys.
+ * @param array|object     $value    The value to traverse into.
+ *
+ * @return mixed
+ */
+function pods_traverse( $traverse, $value ) {
+	if ( ! $traverse && ! is_numeric( $traverse ) ) {
+		return $value;
+	}
+	if ( is_scalar( $value ) ) {
```
Comment (sc0ttkclark) [inline includes/data.php:327]:
```suggestion

	if ( is_scalar( $value ) ) {
```

File: includes/data.php [hunk 13]
```diff
@@ -310,6 +310,43 @@ function pods_trim( $input, $charlist = null, $lr = null ) {
...
+
+	if ( ! isset( $value[ $key ] ) ) {
+		return null;
+	}
+	$value = $value[ $key ];
+
+	if ( $traverse ) {
+		$value = pods_traverse( $traverse, $value );
+	}
+	return $value;
```
Comment (sc0ttkclark) [inline includes/data.php:347]:
```suggestion

	if ( is_object( $value ) ) {
		$value = (array) $value;
	}

	if ( ! is_array( $traverse ) ) {
		$traverse = explode( '.', $traverse );
	}

	$key = array_shift( $traverse );

	if ( ! isset( $value[ $key ] ) ) {
		return null;
	}

	$value = $value[ $key ];

	if ( $traverse ) {
		$value = pods_traverse( $traverse, $value );
	}

	return $value;
```

**Pull request merged**